mephisto/core/task_launcher.py

```diff
@@ -82,14 +94,47 @@ def create_assignments(self) -> None:
                     task_run.sandbox,
                 )
                 self.units.append(Unit(self.db, unit_id))
+                self.unlaunched_units.append(Unit(self.db, unit_id))
+
+    def generate_units(self):
+        """ units generator which checks that only 'max_num_concurrent_units' running at the same time,
+        i.e. in the LAUNCHED or ASSIGNED states """
+        while True:
+            for i in range(len(self.launched_units)):
+                unit = self.launched_units[i]
+                status = unit.get_status()
+                if (
+                    status != AssignmentState.LAUNCHED
+                    and status != AssignmentState.ASSIGNED
+                ):
+                    self.launched_units.pop(i)
```

Comment:
This loop isn't quite doing what you want it to do, and has a chance of running over the end. Take the instance where you have 4 launched units, and the first 2 are being popped. Your list will go through the following steps:
```
i = 0
[<completed unit>, <completed_unit>, <launched unit>, <launched unit>]
           ^
pop unit 0

i = 1
[<completed_unit>, <launched unit>, <launched unit>]
                           ^
# We skipped one of the completed units!

i = 2
[<completed_unit>, <launched unit>, <launched unit>]
                                          ^

i = 3
[<completed_unit>, <launched unit>, <launched unit>]
                                                      ^
# We're out of bounds!!
```

Your original implementation of `launched_units` using `dict`s was actually fine, the only one that needed to change was `unlaunched_units`. It's important to consider whether or not direct indexing is important or useful when making choices on which data structure you need. 

For `unlaunched_units`, order doesn't matter, and it doesn't matter which one we take. Thus a `list` is appropriate as we can always remove from the front for free, and we can create slices for free as well. 
For `launched_units`, we don't know when a unit will be completed, and we don't know from what index we'll need to remove from the `launched_units` data structure. Removing from a random index in a list costs `n`, and can lead to array index bugs as you have here. Removing from a `dict` is constant time. Admittedly you run into an issue where you can't just iterate over the dict's values due to the bounds changing, _but_ you can avoid this by having the iterating loop _find_ the elements to remove (creating a temporary list of `db_id`s to pop), and then iterating over that list immediately after. That prevents you from needing to duplicate the dict every iteration.